engine/gamesys/src/gamesys/scripts/script_resource.cpp

```diff
@@ -838,6 +874,707 @@ static int SetTexture(lua_State* L)
+ *
+ * ```lua
+ * function init(self)
+ *     local params = {
+ *         texture = "/main/my_256x256_texture.texturec",
+ *         animations = {
+ *             {
+ *                 id          = "my_animation",
+ *                 width       = 256,
+ *                 height      = 256,
+ *                 frame_start = 1,
+ *                 frame_end   = 2,
+ *             }
+ *         },
+ *         geometries = {
+ *             {
+ *                 vertices = {
+ *                     0,   0,
+ *                     0,   256,
+ *                     256, 256,
+ *                     256, 0
+ *                 },
+ *                 uvs = {
+ *                     0, 0,
+ *                     0, 256,
+ *                     256, 256,
+ *                     256, 0
+ *                 },
+ *                 indices = { 0,1,2,0,2,3 }
+ *             }
+ *         }
+ *     }
+ *     resource.set_atlas("/main/test.a.texturesetc", params)
+ * end
+ * ```
+ */
+
+static int SetAtlas(lua_State* L)
+{
+    DM_LUA_STACK_CHECK(L, 0);
+
+    dmhash_t path_hash = dmScript::CheckHashOrString(L, 1);
+    void* atlas_res    = CheckResource(L, g_ResourceModule.m_Factory, path_hash, "texturesetc");
+    assert(atlas_res);
+
+    dmGameSystemDDF::TextureSet texture_set_ddf = {};
+    uint32_t frame_index_count                  = 0;
+    uint32_t num_geometries                     = 0;
+    uint32_t num_animations                     = 0;
+    float tex_width                             = 0;
+    float tex_height                            = 0;
+
+    luaL_checktype(L, 2, LUA_TTABLE);
+    lua_pushvalue(L, 2);
+
+    // Set texture resource path
+    lua_getfield(L, -1, "texture");
+    {
+        const char* texture_path         = luaL_checkstring(L, -1);
+        dmhash_t tex_canonical_path_hash = GetCanonicalPathHash(texture_path);
+        void* texture_res                = CheckResource(L, g_ResourceModule.m_Factory, tex_canonical_path_hash, "texturec");
+        assert(texture_res);
+        texture_set_ddf.m_Texture        = texture_path;
+        dmGraphics::HTexture texture     = (dmGraphics::HTexture) texture_res;
+        tex_width                        = dmGraphics::GetTextureWidth(texture);
+        tex_height                       = dmGraphics::GetTextureHeight(texture);
+    }
+    lua_pop(L, 1); // "texture"
+
+    // Note: We do separate passes over the lua state to validate the data in the args table,
+    //       this is because we need to allocate dynamic memory and can't use luaL_check** functions
+    //       since they longjmp away so we can't release the memory..
+    ValidateSetAtlasGeometries(L, &num_geometries);
+    ValidateSetAtlasAnimations(L, &num_animations);
+
+    if (num_animations > 0 && num_geometries == 0)
+    {
+        lua_pop(L, 1);
+        return DM_LUA_ERROR("Setting atlas with animations requires geometry");
+    }
```

Comment:
Not sure if I'm being to defensive in general for this feature, but there are basically no range checks or anything in runtime when using the atlas so if the data is not properly configured the engine will just crash with no feedback on as to why it happened

Reply:
It is a balance act for sure.
I think we may add some range checks I guess where it makes sense.
E.g. indices for geometry, frame numbers for animations.
I think since you've already written a validation code path that touches all the data, then it makes sense to add the range checks.